File: swift/Sources/DiskplanCore/Framing.swift
import Foundation

public let maximumFrameLength = 16 * 1024 * 1024

public enum FrameError: Error, Equatable, CustomStringConvertible {
    case truncatedPrefix(received: Int)
    case oversized(length: Int, maximum: Int)
    case truncatedPayload(expected: Int, received: Int)
    case io(String)

    public var description: String {
        switch self {
        case .truncatedPrefix(let received):
            "truncated frame prefix: received \(received) of 4 bytes"
        case .oversized(let length, let maximum):
            "frame payload length \(length) exceeds maximum \(maximum)"
        case .truncatedPayload(let expected, let received):
            "truncated frame payload: received \(received) of \(expected) bytes"
        case .io(let detail):
            "I/O error: \(detail)"
        }
    }
}

public enum FrameCodec {
    public static func read(from handle: FileHandle) throws -> Data? {
        let prefix = try readExactly(upTo: 4, from: handle)
        if prefix.isEmpty {
            return nil
        }
        guard prefix.count == 4 else {
            throw FrameError.truncatedPrefix(received: prefix.count)
        }
        let length = prefix.reduce(0) { ($0 << 8) | Int($1) }
        guard length <= maximumFrameLength else {
            throw FrameError.oversized(length: length, maximum: maximumFrameLength)
        }
        let payload = try readExactly(upTo: length, from: handle)
        guard payload.count == length else {
            throw FrameError.truncatedPayload(expected: length, received: payload.count)
        }
        return payload
    }

    public static func write(_ payload: Data, to handle: FileHandle) throws {
        guard payload.count <= maximumFrameLength else {
            throw FrameError.oversized(length: payload.count, maximum: maximumFrameLength)
        }
        let length = UInt32(payload.count).bigEndian
        do {
            try withUnsafeBytes(of: length) { bytes in
                try handle.write(contentsOf: bytes)
            }
            try handle.write(contentsOf: payload)
        } catch {
            throw FrameError.io(String(describing: error))
        }
    }

    private static func readExactly(upTo count: Int, from handle: FileHandle) throws -> Data {
        var output = Data()
        output.reserveCapacity(count)
        do {
            while output.count < count {
                guard let chunk = try handle.read(upToCount: count - output.count), !chunk.isEmpty else {
                    break
                }
                output.append(chunk)
            }
            return output
        } catch {
            throw FrameError.io(String(describing: error))
        }
    }
}

File: rust/crates/diskplan-core/src/framing.rs
use std::io::{self, Read, Write};

use thiserror::Error;

pub const MAX_FRAME_LENGTH: usize = 16 * 1024 * 1024;

#[derive(Debug, Error)]
pub enum FrameError {
    #[error("I/O error: {0}")]
    Io(#[from] io::Error),
    #[error("truncated frame prefix: received {received} of 4 bytes")]
    TruncatedPrefix { received: usize },
    #[error("frame payload length {length} exceeds maximum {maximum}")]
    Oversized { length: usize, maximum: usize },
    #[error("truncated frame payload: received {received} of {expected} bytes")]
    TruncatedPayload { expected: usize, received: usize },
}

pub fn read_frame<R: Read>(reader: &mut R) -> Result<Option<Vec<u8>>, FrameError> {
    let mut prefix = [0_u8; 4];
    let prefix_read = read_until_eof(reader, &mut prefix)?;
    if prefix_read == 0 {
        return Ok(None);
    }
    if prefix_read != prefix.len() {
        return Err(FrameError::TruncatedPrefix {
            received: prefix_read,
        });
    }

    let length = u32::from_be_bytes(prefix) as usize;
    if length > MAX_FRAME_LENGTH {
        return Err(FrameError::Oversized {
            length,
            maximum: MAX_FRAME_LENGTH,
        });
    }

    let mut payload = vec![0_u8; length];
    let payload_read = read_until_eof(reader, &mut payload)?;
    if payload_read != length {
        return Err(FrameError::TruncatedPayload {
            expected: length,
            received: payload_read,
        });
    }
    Ok(Some(payload))
}

pub fn write_frame<W: Write>(writer: &mut W, payload: &[u8]) -> Result<(), FrameError> {
    if payload.len() > MAX_FRAME_LENGTH {
        return Err(FrameError::Oversized {
            length: payload.len(),
            maximum: MAX_FRAME_LENGTH,
        });
    }
    writer.write_all(&(payload.len() as u32).to_be_bytes())?;
    writer.write_all(payload)?;
    writer.flush()?;
    Ok(())
}

fn read_until_eof<R: Read>(reader: &mut R, buffer: &mut [u8]) -> io::Result<usize> {
    let mut offset = 0;
    while offset < buffer.len() {
        match reader.read(&mut buffer[offset..]) {
            Ok(0) => break,
            Ok(count) => offset += count,
            Err(error) if error.kind() == io::ErrorKind::Interrupted => continue,
            Err(error) => return Err(error),
        }
    }
    Ok(offset)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clean_eof_is_not_a_truncated_frame() {
        assert!(read_frame(&mut &[][..]).unwrap().is_none());
    }

    #[test]
    fn reports_truncated_prefix_and_payload_separately() {
        assert!(matches!(
            read_frame(&mut &[0_u8, 0, 0][..]),
            Err(FrameError::TruncatedPrefix { received: 3 })
        ));
        assert!(matches!(
            read_frame(&mut &[0_u8, 0, 0, 3, 1, 2][..]),
            Err(FrameError::TruncatedPayload {
                expected: 3,
                received: 2
            })
        ));
    }

    #[test]
    fn rejects_oversized_frame_before_allocating_payload() {
        let length = (MAX_FRAME_LENGTH as u32 + 1).to_be_bytes();
        assert!(matches!(
            read_frame(&mut &length[..]),
            Err(FrameError::Oversized { .. })
        ));
    }

    #[test]
    fn writes_big_endian_length() {
        let mut bytes = Vec::new();
        write_frame(&mut bytes, b"abc").unwrap();
        assert_eq!(bytes, b"\0\0\0\x03abc");
    }
}
